Transformations/Transform/Transform.cpp

```diff
@@ -0,0 +1,340 @@
+        auto rotvec = quat2rotvec(quat);
+        double angle = cv::norm(rotvec);
+        axis = rotvec / cv::norm(rotvec);
+
+        return std::make_tuple(angle, axis);
+    }
+
+    cv::Mat rotx(double rad)
+    {
+        cv::Mat rotmat(3, 3, CV_64FC1, cv::Scalar(0));
+        rotmat = cv::Mat::eye(3, 3, CV_64FC1);
+        rotmat.at<double>(1, 1) = cos(rad);
+        rotmat.at<double>(1, 2) = -sin(rad);
+        rotmat.at<double>(2, 1) = sin(rad);
+        rotmat.at<double>(2, 2) = cos(rad);
+
+        return rotmat;
+    }
+
+    cv::Mat roty(double rad)
+    {
+        cv::Mat rotmat(3, 3, CV_64FC1, cv::Scalar(0));
+        rotmat = cv::Mat::eye(3, 3, CV_64FC1);
+        rotmat.at<double>(0, 0) = cos(rad);
+        rotmat.at<double>(0, 2) = sin(rad);
+        rotmat.at<double>(2, 0) = -sin(rad);
+        rotmat.at<double>(2, 2) = cos(rad);
+
+        return rotmat;
+    }
+
+    cv::Mat rotz(double rad)
+    {
+        cv::Mat rotmat(3, 3, CV_64FC1, cv::Scalar(0));
+        rotmat = cv::Mat::eye(3, 3, CV_64FC1);
+        rotmat.at<double>(0, 0) = cos(rad);
+        rotmat.at<double>(0, 1) = -sin(rad);
+        rotmat.at<double>(1, 0) = sin(rad);
+        rotmat.at<double>(1, 1) = cos(rad);
+
+        return rotmat;
+    }
+
+    cv::Mat euler2rotmat(cv::Vec3d euler, std::string orientation = "zyx")
+    {
+        cv::Mat rotmat(3, 3, CV_64FC1, cv::Scalar(0));
+        rotmat = cv::Mat::eye(3, 3, CV_64FC1);
+        std::map<char, cv::Mat_<double>> R;
+        R.insert(std::make_pair('x', rotx(euler[0])));
+        R.insert(std::make_pair('y', roty(euler[1])));
+        R.insert(std::make_pair('z', rotz(euler[2])));
+        for(char &c : orientation)
+        {
+            if(R.count(c) > 0)
+            {
+                rotmat = rotmat * R.find(c)->second;
+            }
+            else
+            {
+                throw "orientation includes an invalid dimension";
+            }
+        }
+
+        return rotmat;
+    }
+
+    cv::Vec3d rotmat2euler(cv::Mat rotmat, std::string orientation = "zyx")
+    {
+        cv::Vec3d euler;
+
+        double sy =
+            sqrt(rotmat.at<double>(0, 0) * rotmat.at<double>(0, 0) + rotmat.at<double>(1, 0) * rotmat.at<double>(1, 0));
+
+        bool singular = sy < 1e-6; // If
+
+        if(!singular)
+        {
+            euler[0] = atan2(rotmat.at<double>(2, 1), rotmat.at<double>(2, 2));
+            euler[1] = atan2(-rotmat.at<double>(2, 0), sy);
+            euler[2] = atan2(rotmat.at<double>(1, 0), rotmat.at<double>(0, 0));
+        }
+        else
+        {
+            euler[0] = atan2(-rotmat.at<double>(1, 2), rotmat.at<double>(1, 1));
+            euler[1] = atan2(-rotmat.at<double>(2, 0), sy);
+            euler[2] = 0;
+        }
+
+        cv::Vec3d euler360 = euler * 180 / CV_PI;
+        std::cout << orientation << "(degrees):\n" << euler360 << std::endl;
+        std::cout << orientation << "(radians):\n" << euler << std::endl;
+
+        return euler;
```

Comment:
Which order will it return in? theta_x theta_y theta_z or theta_z theta_y theta_x - I assume we are talking about zy'x'' / xyz

Reply:
Yes, this function is an inverse of calling `euler2rotmat(euler, 'zyx', INTRINSIC)`. (given that we add Intrinsic/Extrinsic to `euler2rotmat()`. 
As I wrote in my comment to this function, I primarily copied it from https://www.learnopencv.com/rotation-matrix-to-euler-angles/. I'm not comfortable about it. It only supports one format. I need to draw the math on paper in order to comfortably write out the function for all permutations.